localstack/services/s3/provider_native.py

```diff
@@ -0,0 +1,1353 @@
+                response[f"Checksum{checksum_algorithm.upper()}"] = checksum  # noqa
+
+        if range_header := request.get("Range"):
+            range_data = parse_range_header(range_header, s3_object.size)
+            response["ContentLength"] = range_data.content_length
+
+        if s3_object.parts:
+            response["PartsCount"] = len(s3_object.parts)
+
+        if s3_object.version_id:
+            response["VersionId"] = s3_object.version_id
+
+        if s3_object.website_redirect_location:
+            response["WebsiteRedirectLocation"] = s3_object.website_redirect_location
+
+        # TODO: missing return fields:
+        # Expiration: Optional[Expiration]
+        # Restore: Optional[Restore]
+        # ArchiveStatus: Optional[ArchiveStatus]
+
+        # RequestCharged: Optional[RequestCharged]
+        # ReplicationStatus: Optional[ReplicationStatus]
+        # ObjectLockMode: Optional[ObjectLockMode]
+        # ObjectLockRetainUntilDate: Optional[ObjectLockRetainUntilDate]
+        # ObjectLockLegalHoldStatus: Optional[ObjectLockLegalHoldStatus]
+
+        return response
+
+    def delete_object(
+        self,
+        context: RequestContext,
+        bucket: BucketName,
+        key: ObjectKey,
+        mfa: MFA = None,
+        version_id: ObjectVersionId = None,
+        request_payer: RequestPayer = None,
+        bypass_governance_retention: BypassGovernanceRetention = None,
+        expected_bucket_owner: AccountId = None,
+    ) -> DeleteObjectOutput:
+        # TODO: implement bypass_governance_retention, it is done in moto
+        store = self.get_store(context.account_id, context.region)
+        if not (s3_bucket := store.buckets.get(bucket)):
+            raise NoSuchBucket("The specified bucket does not exist", BucketName=bucket)
+
+        if s3_bucket.versioning_status is None:
+            if version_id and version_id != "null":
+                raise InvalidArgument(
+                    "Invalid version id specified",
+                    ArgumentName="versionId",
+                    ArgumentValue=version_id,
+                )
+
+            found_object = s3_bucket.objects.pop(key, None)
+            # TODO: RequestCharged
+            if found_object:
+                self._storage_backend.delete_key_fileobj(bucket_name=bucket, object_key=key)
+            return DeleteObjectOutput()
+
+        if not version_id:
+            delete_marker_id = generate_version_id(s3_bucket.versioning_status)
+            delete_marker = S3DeleteMarker(key=key, version_id=delete_marker_id)
+            # TODO: verify with Suspended bucket? does it override last version or still append?? big question
+            #  I think it puts a delete marker with a `null` VersionId, which deletes the object under
+            s3_bucket.objects.set(key, delete_marker)
+            return DeleteObjectOutput(VersionId=delete_marker.version_id, DeleteMarker=True)
+
+        if key not in s3_bucket.objects:
```

Comment:
issue: KeyStore and VersionedKeyStore do not implement `__contains__` yet but this will be fixed in the next PR 😬 hard to fully test everything, this will be fixed as I can test more and more with more things implemented (implementing `PutBucketVersioning` allows me to test much more already and will contains all these small fixes).